lib/spack/docs/packaging_guide.rst

```diff
@@ -2016,45 +2016,124 @@ the Python extensions provided by them: once for ``+python`` and once
 for ``~python``.  Other than using a little extra disk space, that
 solution has no serious problems.
 
------------------------------------
-Implementing the ``install`` method
------------------------------------
+.. _installation_procedure:
 
-The last element of a package is its ``install()`` method.  This is
+---------------------------------------
+Implementing the installation procedure
+---------------------------------------
+
+The last element of a package is its **installation procedure**.  This is
 where the real work of installation happens, and it's the main part of
 the package you'll need to customize for each piece of software.
 
-.. literalinclude::  ../../../var/spack/repos/builtin/packages/mpfr/package.py
-   :pyobject: Mpfr.install
-   :linenos:
+Defining an installation procedure means overriding a set of methods or attributes
+that will be called at some point during the installation of the package.
+What determines the actual set of entities that are available for overriding
+is the package base class, which is usually specialized for a given build system.
+The classes that are currently provided by Spack are:
+
+    +------------------------------------+----------------------------------+
+    |                                    |   **Base class purpose**         |
+    +====================================+==================================+
+    |          :py:class:`.Package`      | General base class not           |
+    |                                    | specialized for any build system |
+    +------------------------------------+----------------------------------+
+    |   :py:class:`.MakefilePackage`     | Specialized class for packages   |
+    |                                    | that are built invoking an       |
+    |                                    | hand-written Makefile            |
+    +------------------------------------+----------------------------------+
+    |   :py:class:`.AutotoolsPackage`    | Specialized class for packages   |
+    |                                    | that are built using GNU         |
+    |                                    | Autotools                        |
+    +------------------------------------+----------------------------------+
+    |  :py:class:`.CMakePackage`         | Specialized class for packages   |
+    |                                    | that are built using CMake       |
+    +------------------------------------+----------------------------------+
+    |  :py:class:`.RPackage`             | Specialized class for            |
+    |                                    | :py:class:`.R` extensions        |
+    +------------------------------------+----------------------------------+
```

Comment:
The docs should be consistent for R (and Python) too... that the package superclass depends on the *BUILD SYSTEM* being used, not the language of the package.  For example, a Python extension installed with CMake (they do exist) would `extends('python')`, but it would also subclass from `CMakePackage`.

Similarly, some upstream packages come with a "configure" script that is not a real Autotools configure script.  Although this duck has the same name, it is not really a duck: it can't be regenerated with `autoreconf`, its command line options may be different from those of a standard Autotools-built `configure`, etc.  Therefore, such packages need to subclass `Packagd`.

`PythonPackage` should be added to this list.